Lib/fontTools/pens/ttGlyphPen.py

```diff
@@ -103,72 +202,79 @@ def closePath(self):
-            component.flags = componentFlags
-            components.append(component)
-        return components
 
-    def glyph(self, componentFlags=0x4):
-        """Returns a :py:class:`~._g_l_y_f.Glyph` object representing the glyph."""
-        assert self._isClosed(), "Didn't close last contour."
+class TTGlyphPointPen(TTGlyphBasePen, LogMixin, AbstractPointPen):
+    """
+    Pen used for drawing to a TrueType glyph.
 
-        components = self._buildComponents(componentFlags)
+    This pen can be used to construct or modify glyphs in a TrueType format
+    font. After using the pen to draw points, use the ``.glyph()`` method to
+    retrieve a :py:class:`~._g_l_y_f.Glyph` object representing the glyph.
+    """
 
-        glyph = Glyph()
-        glyph.coordinates = GlyphCoordinates(self.points)
-        glyph.coordinates.toInt()
-        glyph.endPtsOfContours = self.endPts
-        glyph.flags = array("B", self.types)
-        self.init()
+    def _decompose(
+        self,
+        glyphName: str,
+        transformation: Tuple[float, float, float, float, float, float],
+    ):
+        tpen = TransformPointPen(self, transformation)
+        self.glyphSet[glyphName].drawPoints(tpen)
+
+    def beginPath(
+        self, identifier: Optional[str] = None, **kwargs: Any
+    ) -> None:
+        """
+        Start a new sub path.
+        """
+        assert self.currentPath is None
+        self.currentPath = []
 
-        if components:
-            glyph.components = components
-            glyph.numberOfContours = -1
+    def endPath(self) -> None:
+        """
+        End the current sub path.
+        """
+        # TrueType contours are always "closed"
+        assert self.currentPath is not None
+        if self.currentPath:
+            endPt = len(self.points) - 1
+            self.endPts.append(endPt)
+        self.currentPath = None
+
+    def addPoint(
+        self,
+        pt: Tuple[float, float],
+        segmentType: Optional[str] = None,
+        smooth: bool = False,
+        name: Optional[str] = None,
+        identifier: Optional[str] = None,
+        **kwargs: Any
+    ) -> None:
+        """
+        Add a point to the current sub path.
+        """
+        assert self.currentPath is not None
+        if segmentType is None:
+            self.types.append(0)  # offcurve
+        elif segmentType in ("qcurve", "line"):
+            self.types.append(1)  # oncurve
         else:
-            glyph.numberOfContours = len(glyph.endPtsOfContours)
-            glyph.program = ttProgram.Program()
-            glyph.program.fromBytecode(b"")
+            # cubic curves are not supported
+            raise NotImplementedError
 
-        return glyph
+        self.currentPath.append(pt)
+        self.points.append(pt)
+
+    def glyph(self, componentFlags: int = 0x4) -> Glyph:
+        """
+        Returns a :py:class:`~._g_l_y_f.Glyph` object representing the glyph.
+        """
+        assert self.currentPath is None
```

Comment:
`assert self.currentPath is None` could be the content of `TTGlyphPointPen._isClosed`, the point-pen version of `_isClosed` method of `TTGlyphPen`. The base class could define an "abstract" `_isClosed` that raises NotImplementedError, so the subclasses must override it with specific implementations. Then you would no longer need to have separate `_glyph` and `glyph` methods, with the latter only difference in the way to check that the countor is closed.

Reply:
About `self.currentPath`: why is it even a list? It's effectively a three-state flag: 
1. no current path (not a valid contour)
2. there is a current path but no points have been added (not a valid contour)
3. there is a current path with at least one point (a valid contour)

Idea: call it `_currentContourStartIndex` or something, set to `None` upon reset, and set to `len(self.points)` when starting a new contour. Then you can compare with `len(self.points)` to see if there are any points.